File: src/Niche/Prolog/builder.pl
text_builder(builder(text)).
html_builder(builder(html)).

add_title(builder(text), Title, Parts, Result) :-
    atom_concat('# ', Title, Line),
    append(Parts, [Line], Result).
add_title(builder(html), Title, Parts, Result) :-
    atomic_list_concat(['<h1>', Title, '</h1>'], Line),
    append(Parts, [Line], Result).

add_section(builder(text), Heading, Body, Parts, Result) :-
    atom_concat('## ', Heading, Line),
    append(Parts, [Line, Body], Result).
add_section(builder(html), Heading, Body, Parts, Result) :-
    atomic_list_concat(['<h2>', Heading, '</h2>'], H),
    atomic_list_concat(['<p>', Body, '</p>'], B),
    append(Parts, [H, B], Result).

build(builder(text), Parts, Result) :- atomic_list_concat(Parts, '\n', Result).
build(builder(html), Parts, Result) :- atomic_list_concat(Parts, '', Result).

build_availability_report(Builder, Result) :-
    add_title(Builder, 'Service status', [], P1),
    add_section(Builder, 'Availability', '99.95%', P1, P2),
    build(Builder, P2, Result).

run :-
    text_builder(Text),
    html_builder(Html),
    build_availability_report(Text, TextResult),
    build_availability_report(Html, HtmlResult),
    format('~w~n---~n~w~n', [TextResult, HtmlResult]).

File: src/Enterprise/Delphi/Builder.pas
program Builder;

{$APPTYPE CONSOLE}

uses
  SysUtils, Classes;

type
  IReportBuilder = interface
    ['{CE195E78-C992-45B9-AEAF-A57ED9547EC6}']
    procedure Reset;
    procedure AddTitle(const ATitle: string);
    procedure AddSection(const AHeading, ABody: string);
    function Build: string;
  end;

  TTextReportBuilder = class(TInterfacedObject, IReportBuilder)
  protected
    FParts: TStringList;
  public
    constructor Create;
    destructor Destroy; override;
    procedure Reset;
    procedure AddTitle(const ATitle: string); virtual;
    procedure AddSection(const AHeading, ABody: string); virtual;
    function Build: string;
  end;

  THtmlReportBuilder = class(TTextReportBuilder)
  public
    procedure AddTitle(const ATitle: string); override;
    procedure AddSection(const AHeading, ABody: string); override;
  end;

constructor TTextReportBuilder.Create;
begin
  inherited;
  FParts := TStringList.Create;
end;

destructor TTextReportBuilder.Destroy;
begin
  FParts.Free;
  inherited;
end;

procedure TTextReportBuilder.Reset;
begin
  FParts.Clear;
end;

procedure TTextReportBuilder.AddTitle(const ATitle: string);
begin
  FParts.Add('# ' + ATitle);
end;

procedure TTextReportBuilder.AddSection(const AHeading, ABody: string);
begin
  FParts.Add('## ' + AHeading);
  FParts.Add(ABody);
end;

function TTextReportBuilder.Build: string;
begin
  Result := TrimRight(FParts.Text);
end;

procedure THtmlReportBuilder.AddTitle(const ATitle: string);
begin
  FParts.Add('<h1>' + ATitle + '</h1>');
end;

procedure THtmlReportBuilder.AddSection(const AHeading, ABody: string);
begin
  FParts.Add('<h2>' + AHeading + '</h2><p>' + ABody + '</p>');
end;

function BuildAvailabilityReport(const ABuilder: IReportBuilder): string;
begin
  ABuilder.Reset;
  ABuilder.AddTitle('Service status');
  ABuilder.AddSection('Availability', '99.95%');
  Result := ABuilder.Build;
end;

begin
  Writeln(BuildAvailabilityReport(TTextReportBuilder.Create));
  Writeln('---');
  Writeln(BuildAvailabilityReport(THtmlReportBuilder.Create));
end.
